kubed-core/src/main/kotlin/kubed/scale/ContinuousScale.kt

```diff
@@ -1,54 +1,71 @@
 package kubed.scale
 
 import kubed.array.bisect
+import kubed.interpolate.*
 import java.util.*
 import kotlin.comparisons.naturalOrder
 import kotlin.math.min
 
-abstract class ContinuousScale<R>(val interpolate: (R, R) -> (Double) -> R,
-                                  val uninterpolate: ((R, R) -> (R) -> Double)? = null,
-                                  val rangeComparator: Comparator<R>? = null) : Scale<Double, R> {
+abstract class ContinuousScale<R>(
+    val reinterpolatorFactory: ReinterpolatorFactory<R>,
+    val deinterpolatorFactory: DeinterpolatorFactory<R>? = null,
+    val rangeComparator: Comparator<R>? = null
+) : Scale<Double, R> {
+
     override val domain: MutableList<Double> = mutableListOf(0.0, 1.0)
     override val range: MutableList<R> = mutableListOf()
-    var clamp: Boolean = false
-
-    var piecewiseOutput: ((List<Double>, List<R>,
-                           (Double, Double) -> (Double) -> Double,
-                           (R, R) -> (Double) -> R) -> (Double) -> R)? = null
-    var output: ((Double) -> R)? = null
 
-    var piecewiseInput: ((List<Double>, List<R>,
-                         (Double, Double) -> (Double) -> Double,
-                         (R, R) -> (R) -> Double) -> (R) -> Double)? = null
-    var input: ((R) -> Double)? = null
-
-    protected abstract fun deinterpolate(a: Double, b: Double): (Double) -> Double
-
-    protected abstract fun reinterpolate(a: Double, b: Double): (Double) -> Double
+    var clamp: Boolean = false
 
-    protected fun deinterpolateClamp(deinterpolate: (Double, Double) -> (Double) -> Double): (Double, Double) -> (Double) -> Double {
-        return fun(a: Double, b: Double): (Double) -> Double {
-            val d = deinterpolate(a, b)
-            return fun(x: Double): Double {
-                return when {
+    var piecewiseReinterpolatorFactory: PiecewiseReinterpolatorFactory<R>? = null
+    var piecewiseReinterpolator: Reinterpolator<R>? = null
+
+    var piecewiseDeinterpolatorFactory: PiecewiseDeinterpolatorFactory<R>? = null
+    var piecewiseDeinterpolator: Deinterpolator<R>? = null
+
+    /**
+     * Creates an interpolator function from t in [0, 1] to x in [a, b].
+     * Note: if t is not in [0, 1], x can be outside [a, b].
+     */
+    protected abstract fun deinterpolatorOf(a: Double, b: Double): Deinterpolator<Double>
```

Comment:
Sadly, `Deinterpolator<Double>` and `Reinterpolator<Double>` are still interchangeable from the compiler's point of view, since both are `(Double, Double) -> (Double) -> Double`, but at least the developer can now more easily tell. I wish Kotlin supported distinct classes, not just typealiases.

Reply:
Oh, meant to say `DeinterpolatorFactory<Double>`, `ReinterpolatorFactory<Double>`, but it's true for `Deinterpolator<Double>` and `Reinterpolator<Double>` as well.

Reply:
We could always create:

```kotlin
interface DeinterpolatorFactory {
    fun create(a: Double, b: Double)
}
```
and 
```kotlin
interface ReinterpolatorFactory {
    fun create(a: Double, b: Double)
}
```

If type safety was really a concern.

Reply:
Yeah, I experimented with interfaces and found them to be somewhat less ergonomic.

Reply:
Yeah I experimented with it before and didn’t care for it either